examples/src/app/tree-crdt/TreeView.tsx

```diff
@@ -0,0 +1,222 @@
+
+  const [mousePosition, setMousePosition] = useState<Point | null>(null)
+
+  let nodes = useMemo(
+    () => new NodeBox(props.root).getChildren(),
+    [props.root, props.root.tree.structure],
+  )
+  const height = 600
+  const width = 600
+  const xScale = (x: number) => x * width
+  const yScale = (y: number) => y * height
+
+  const scalePoint = (point: Point) => ({
+    x: xScale(point.x),
+    y: yScale(point.y),
+    radius: point.radius,
+  })
+
+  let tentativeEdgePath: [Point, Point] | null = null
+  if (dragOrigin !== null) {
+    if (dragTarget !== null) {
+      tentativeEdgePath = [scalePoint(dragOrigin), scalePoint(dragTarget)]
+    } else {
+      if (mousePosition && dragOrigin) {
+        tentativeEdgePath = [scalePoint(dragOrigin), mousePosition]
+      }
+    }
+  }
+
+  return (
+    <svg
+      style={{ width, height }}
+      onMouseUp={mouseUp}
+      onMouseMove={(e) => {
+        if (e.buttons === 1) {
+          let rect = e.currentTarget.getBoundingClientRect()
+          let x = e.clientX - rect.left
+          let y = e.clientY - rect.top
+
+          setMousePosition({ x, y })
+        }
+      }}
+    >
+      {nodes.map(
+        (node) =>
+          node.parent && (
+            <Arrow
+              key={node.treeNode.id()}
+              origin={scalePoint(node)}
+              target={scalePoint(node.parent)}
+              animate={true}
+            />
+          ),
+      )}
+
+      {tentativeEdgePath && (
+        <Arrow origin={tentativeEdgePath[0]} target={tentativeEdgePath[1]} animate={false} />
+      )}
+
+      {nodes.map((node) => (
+        <g
+          key={node.treeNode.id()}
+          style={{
+            transform: `translate(${xScale(node.x)}px, ${yScale(node.y)}px)`,
+            transition: 'transform 0.2s',
+            animation: 'fadeInAnimation ease 0.2s',
+          }}
+          onMouseMove={(e) => {
+            if (e.buttons === 1) {
+              mouseOver(node)
+            }
+          }}
+          onMouseOut={() => mouseOver(null)}
+          onDoubleClick={() => {
+            node.treeNode.addChild()
+          }}
+        >
+          <circle
+            r={node.radius}
+            fill={node.treeNode.color()}
+            stroke={node === dragOrigin || node === dragTarget ? 'white' : 'black'}
+            strokeWidth={2}
+          />
+          <text
+            textAnchor="middle"
+            dominantBaseline="middle"
+            style={{ fontSize: 12, userSelect: 'none', fontFamily: 'monospace' }}
+          >
+            {node.treeNode.id() === ROOT_ID ? 'root' : node.treeNode.id()}
+          </text>
+        </g>
+      ))}
+    </svg>
+  )
+}
+
+export function TreeView() {
+  const [tree, setTree] = useState<YTreeNode | null>(null)
+  const [_, bumpTreeVersion] = useState(0)
+  const treeMap = useMap<Y.Map<any>>('tree')
+
+  useEffect(() => {
+    let tree = new YTree(treeMap)
+    tree.setOnChange(() => {
+      bumpTreeVersion((v) => v + 1)
+    })
```

Comment:
Do you need to unsubscribe this `onChange` callback on unmount?

Reply:
I think it's probably ok not to here because the lifetime of `bumpTreeVersion` and `tree` should always be the same as the life of the component, but it doesn't hurt to so I've done it.